core/control-panel/impl/src/services/registry.rs

```diff
@@ -48,14 +58,146 @@ impl RegistryService {
+                    };
+
+                    where_clause = where_clause.to_owned().and_fullname(&fullname);
+                }
+                SearchRegistryFilterKindDTO::Kind(kind) => {
+                    where_clause = where_clause.to_owned().and_kind(kind.into());
+                }
+            });
+
+        let entry_ids = self.registry_repository.find_ids_where(where_clause, None);
+        let paginated_ids = paginated_items(PaginatedItemsArgs {
+            offset: search.pagination.to_owned().and_then(|p| p.offset),
+            limit: search.pagination.and_then(|p| p.limit),
+            default_limit: Some(Self::DEFAULT_SEARCH_LIMIT),
+            max_limit: Some(Self::MAX_SEARCH_LIMIT),
+            items: &entry_ids,
+        })?;
+
+        Ok(PaginatedData {
+            total: paginated_ids.total,
+            next_offset: paginated_ids.next_offset,
+            items: paginated_ids
+                .items
+                .into_iter()
+                .flat_map(|id| match self.get(&id) {
+                    Ok(entry) => Some(entry),
+                    Err(_) => None,
+                })
+                .collect::<Vec<RegistryEntry>>(),
+        })
+    }
+
     /// Creates a new registry entry and returns it.
-    pub fn create(&self) -> ServiceResult<RegistryEntry> {
-        unimplemented!()
+    pub fn create(&self, input: RegistryEntryInput) -> ServiceResult<RegistryEntry> {
+        let mut entry = RegistryEntry::new();
+
+        RegistryMapper::fill_from_create_input(&mut entry, &input);
+
+        match &input.value {
+            control_panel_api::RegistryEntryValueInput::WasmModule(module) => {
+                let artifact_id = self.artifact_service.create(module.wasm_module.clone())?;
+
+                entry.value = RegistryValue::WasmModule(WasmModuleRegistryValue {
+                    wasm_artifact_id: artifact_id,
+                    version: module.version.clone(),
+                    dependencies: module
+                        .dependencies
+                        .iter()
+                        .map(|dep| dep.clone().into())
+                        .collect(),
+                });
+            }
+        };
+
+        // There can only be one latest entry, so we need to remove the latest tag from the previous entry if it exists.
+        if entry.tags.contains(&LATEST_TAG.to_string()) {
+            let previous_latest = self.registry_repository.find_ids_where(
+                RegistryWhere::clause()
+                    .and_fullname(&entry.fullname())
+                    .and_tags(vec![LATEST_TAG.to_string()]),
+                None,
+            );
+
+            previous_latest
+                .iter()
+                .filter_map(|id| self.get(id).ok())
+                .for_each(|mut e| {
+                    e.tags.retain(|t| t != LATEST_TAG);
+                    self.registry_repository.insert(e.id, e);
+                });
+        }
+
+        entry.validate()?;
```

Comment:
remove artifact if validation failes?

Reply:
also, the `latest` tag could be lost if validation fails.

Reply:
Great point, since we're not adding a `panic` there, just made a change that accounts for that